src/java/org/apache/cassandra/db/compaction/CompactionManager.java

```diff
@@ -680,12 +680,28 @@ public AllSSTableOpStatus performGarbageCollection(final ColumnFamilyStore cfSto
             @Override
             public Iterable<SSTableReader> filterSSTables(LifecycleTransaction transaction)
             {
-                Iterable<SSTableReader> originals = transaction.originals();
+                List<SSTableReader> sstables = Lists.newArrayList(transaction.originals());
+                Iterator<SSTableReader> iter = sstables.iterator();
                 if (cfStore.getCompactionStrategyManager().onlyPurgeRepairedTombstones())
-                    originals = Iterables.filter(originals, SSTableReader::isRepaired);
-                List<SSTableReader> sortedSSTables = Lists.newArrayList(originals);
-                Collections.sort(sortedSSTables, SSTableReader.maxTimestampAscending);
-                return sortedSSTables;
+                {
+                    while (iter.hasNext())
+                    {
+                        SSTableReader sstable = iter.next();
+                        if (!sstable.isRepaired())
+                        {
+                            try
+                            {
+                                transaction.cancel(sstable);
+                            }
+                            finally
+                            {
+                                iter.remove();
```

Comment:
Removal in the interior of the array list makes this an O(n^2) operation, which can be a problem when the number of sstables can grow proportionally with data size (e.g. with LCS and UCS).
I'd turn this into an iteration over the source, deciding to add to an initially empty array or cancel.